src/metrics.h

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+#ifndef METADATA_AGENT_METRICS_H_
+#define METADATA_AGENT_METRICS_H_
+
+#include <opencensus/stats/stats.h>
+#include <string>
+
+namespace google {
+
+class Metrics {
+ public:
+  static const char kGceApiRequestErrors[];
+
+  // Record an API request error on a method, e.g. oauth2 is one of the methods.
+  static void RecordGceApiRequestErrors(int64_t value, const std::string& method);
+
+  // Serialize all the available metrics as prometheus text format.
+  static std::string SerializeMetricsToPrometheusTextFormat();
+
+  // View Descriptor accessors. If the view descriptor variable is not
+  // initialized, these methods will initialize the variable.
+  static const ::opencensus::stats::ViewDescriptor
```

Comment:
Can this be private?

Reply:
no. I tried to implement `GetGceApiRequestErrorsCumulativeViewIntData`, but it the data cannot propogated to this view because 1. the view needs to be created before the metric is recorded, and 2. for testing, we need to call `TestUtils::Flush()` to propagate to view. If this functino is private, it cannot be used in `oauth2_unittest`. We may put it as private function and make a class in `oauth2_unittest` as friend class, but if we add more metrics across files, you will friend a couple more - so I am not sure that's what we want.

In order to create a view from the view descriptor, and run `TestUtils::Flush()`, I put this function as public function.

Reply:
Also note that the standard trick for accessing private functions in tests is to mark them `protected` in the implementation class, and add a subclass in the test that makes them `public`. This can be made to work on non-virtual and static functions as well.